InvokeEffectorOnCollectionSensorChange

sjcorbett opened the pull request:
Tracks a collection and invokes effectors when elements are added and removed.

neykov commented on core/src/main/java/org/apache/brooklyn/policy/InvokeEffectorOnSetChange.java:
```diff
@@ -0,0 +1,203 @@
+    @Override
+    public void setEntity(EntityLocal entity) {
+        super.setEntity(entity);
+        Sensor<? extends Collection<?>> sensor =
```
Worth a comment that the method will be executed on rebind. In case of adding additional validation should be backwards compatible with existing persistence state.

sjcorbett commented on core/src/main/java/org/apache/brooklyn/policy/InvokeEffectorOnSetChange.java:
```diff
@@ -0,0 +1,203 @@
+        synchronized (updateLock) {
+            // Not using .immutableCopy() in case either set contains `null`.
+            Sets.difference(newValue, previous).copyInto(added);
+            Sets.difference(previous, newValue).copyInto(removed);
```
We have to take the difference with the lock so that we can determine whether the collection is changed and thus should set the `previous` sensor. I'll move the `onAdded` and `onRemoved` calls out of the critical section.

sjcorbett replied:
Though to be honest since effectors are asynchronous this will just be a problem further down the line, won't it?

sjcorbett commented on core/src/main/java/org/apache/brooklyn/policy/InvokeEffectorOnSetChange.java:
```diff
@@ -0,0 +1,203 @@
+                entity.sensors().set(getPreviousSensor(), previous);
+                requestPersist();
+                // In advance of the next event.
+                previous = Collections.unmodifiableSet(newValue);
```
Just to be clear - do you mean to ask whether we really need a separate _field_? Rather than sensor. Assuming so, yes, the previous sensor is just for users to look at. I thought we'd decided that was desirable previously. I don't have any strong feelings about the `previous` sensor - am happy enough to get rid of it but can see that it might be useful.

sjcorbett commented:
@neykov thank you for your comments. I've addressed everything I didn't comment on. I've also renamed the policy to `InvokeEffectorOnCollectionSensorChange` to be consistent with the existing `InvokeEffectorOnSensorChange`, but no strong feelings there really.

sjcorbett commented on core/src/main/java/org/apache/brooklyn/policy/InvokeEffectorOnSetChange.java:
```diff
@@ -0,0 +1,203 @@
+    @Override
+    public void onEvent(SensorEvent<Collection<?>> event) {
+        final Set<Object> added = new HashSet<>(), removed = new HashSet<>();
+        Set<Object> newValue = new HashSet<>(event.getValue());
```
Is there a way to get GitHub to hide out of date diffs now?

sjcorbett commented:
I will merge this.
